### DISCUSSION TITLE
Can't start midi mode for Coldtype

### DISCUSSION THREAD

**Discussion by jesi-rgb** [coldtype/coldtype#65] [Q&A] [answered]
Hello everyone!

I want to use my little Novation Launchkey as a midi device with Coldtype.

As stated in the tutorial:

> With that in mind, I’ll use the Launch Control as an example. So with that device plugged in, when I run `coldtype -mi` in my coldtype virtualenv...

supposedly I should be able all the midi devices I have connected, which is only this one.

But instead, I just get an error:

```python
➜ coldtype -mi
usage: coldtype [-h] [-v] [-sv] [-r {drawbot,cairo,svg,skia,pickle}] [-cpu] [-s SCALE]
                [-a] [-b] [-rls] [-mm] [-c CONFIG] [-p PROFILE] [-fmt FORMAT]
                [-i INDICES] [-of OUTPUT_FOLDER] [-wpass WINDOW_PASSTHROUGH]
                [-wt WINDOW_TRANSPARENT] [-wc WINDOW_CHROMELESS]
                [-wb WINDOW_BACKGROUND] [-wf WINDOW_FLOAT] [-wo WINDOW_OPACITY]
                [-wp WINDOW_PIN] [-wpi WINDOW_PIN_INSET] [-wcs WINDOW_CONTENT_SCALE]
                [-mn MONITOR_NAME] [-ec EDITOR_COMMAND] [-minc MANY_INCREMENT]
                [-ps PREVIEW_SCALE] [-fd FONT_DIRS] [-ff FUNCTION_FILTERS]
                [-tc THREAD_COUNT] [-mp MULTIPLEX] [-dt DEBOUNCE_TIME]
                [-rd REFRESH_DELAY] [-in INLINE_FILES] [-bw BLENDER_WATCH]
                [-bap BLENDER_APP_PATH] [-wv WEBVIEWER] [-ws WEBSOCKET]
                [-wvp WEBVIEWER_PORT] [-wsp WEBSOCKET_PORT] [-nv NO_VIEWER]
                [-nm NO_MIDI] [-mi MIDI_INFO] [-ns NO_SOUND] [-nve NO_VIEWER_ERRORS]
                [-ea ENABLE_AUDIO]
                [file] [inputs ...]
coldtype: error: argument -mi/--midi-info: expected one argument
```

Must I specify a file like in `coldtype test.py` or am I missing something?

If that's the case, running `coldtype -mi test.py` just runs test.py as it usually would, but no midi devices are shown, let alone the logging of the keys when pressed.

I really want to use this tool, I am very excited about it, but can't get it to work :(

Any help would be much appreciated. Thanks a lot for your precious time beforehand.

**Comment by stenson** [comment, marked as answer]
Hi @jesi-rgb — thanks for trying out Coldtype, and apologies for the confusion!

This is my bad — I changed the flag to be `-mi 1` instead of just `-mi` a couple months ago, but I didn't update the documentation.

Basically, you just always need to pass a "truthy" value (like 1) to a flag in order to turn it on, so here `coldtype -mi 1` should work and print out your midi info.

Also just updated here: https://coldtype.goodhertz.com/tutorials/midi.html

**Comment by jesi-rgb** [comment]
Thanks for your kind response, @stenson! Really worked.

On the same topic of "_maybe not updated docs_", I am trying the Generic midi configuration, but it doesn't work either.

Using `coldtype -mi 1` now succesfully logs the devices and their activity. This is an example with my own device:

```python
Launchkey Mini MK3 MIDI Port <CONTROLLER: 28, value: 74, channel: 1>
Launchkey Mini MK3 MIDI Port <CONTROLLER: 28, value: 73, channel: 1>
Launchkey Mini MK3 MIDI Port <CONTROLLER: 28, value: 72, channel: 1>
Launchkey Mini MK3 MIDI Port <CONTROLLER: 28, value: 71, channel: 1>
Launchkey Mini MK3 MIDI Port <CONTROLLER: 28, value: 70, channel: 1>
Launchkey Mini MK3 MIDI Port <CONTROLLER: 28, value: 69, channel: 1>
Launchkey Mini MK3 MIDI Port <CONTROLLER: 28, value: 68, channel: 1>
```

I want to target that controller, n 28, which is a knob.

Using the code in the tutorial, I run into an error saying that Generic is missing a channel value. Because of the logs with the keyboard showing up `channel: 1`, I assume that is the channel I'm targetting too. So I add the name of the midi device, the controller I am targetting (28) and the channel in which it works. So the snippet of code changes like this:
```diff
@renderable(rstate=1)
def use_midi(r, rs):
-    controller = Generic("Launchkey Mini MK3 MIDI Port", rs.midi)
+    controller = Generic("Launchkey Mini MK3 MIDI Port", rs.midi, channel=1)
    fader = controller(28, 0.5)  # returns a value between 0 and 1
    return DATPen().oval(r.take(fader, "mdx").square()).f(hsl(0.65))
```

Great! Now, there is no error, but it simply does not work. Am I missing something again? Maybe the `mdx` value is different? What do you think?


I am sorry for being so annoying, but I've never been more excited about some python code since I tried Manim and I really want to experiment with this tool. Thanks a lot for your help, Rob! Really appreciate it.

PS.: Also, I am not that experienced at python, but I can try my best to help develop something, at least updating the docs. I also want to support this tools in some way.

**Reply by stenson** [reply]
Yikes, yeah I have to admit I haven't used the MIDI functionality for a while, so the docs have definitely aged a little — sorry about that!

You've figured out the channel issue completely, but yes, another change from a little while ago, basically anything that can change appearance over time needs to be decorated as an `@animation` now (as opposed to a `@renderable`, which is what the example in the docs was). I've updated the docs again, but here's the change you'd need to make:

```python
from coldtype import *
from coldtype.midi.controllers import Generic

@animation(rstate=1)
def use_midi(f, rs):
    controller = Generic("Launch Control XL", rs.midi, channel=9)
    fader = controller(77, 0.5) # returns a value between 0 and 1
    return (DATPen()
        .oval(f.a.r.take(fader, "mdx").square())
        .f(hsl(0.65)))
```

**Reply by jesi-rgb** [reply]
Oh! I see. Got it! I enjoyed a lot the animation part of what I was able to figure out so far. Quite addictive 😈

Will try the animation immediately, I am so excited!!!

Thanks a lot again, Rob! Very kind of you. I am so happy that something like this exists, if you need anything please let me know.
